CMake/mlpack.cmake

```
@@ -0,0 +1,682 @@
+      find_package(LAPACK QUIET)
+    endif()
+    get_deps(https://files.mlpack.org/armadillo-${ARMADILLO_VERSION}.tar.gz armadillo armadillo-${ARMADILLO_VERSION}.tar.gz)
+    set(ARMADILLO_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    if (NOT CMAKE_CROSSCOMPILING)
+      find_armadillo(${CMAKE_BINARY_DIR})
+    endif()
+    # Include directories for the previous dependencies.
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${ARMADILLO_INCLUDE_DIRS})
+    set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${ARMADILLO_LIBRARIES})
+  endif()
+
+  find_ensmallen(${CMAKE_BINARY_DIR})
+  if (NOT ENSMALLEN_FOUND)
+    get_deps(https://www.ensmallen.org/files/ensmallen-latest.tar.gz ensmallen ensmallen-latest.tar.gz)
+    set(ENSMALLEN_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    find_ensmallen(${CMAKE_BINARY_DIR})
+    if (ENSMALLEN_FOUND)
+      set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${ENSMALLEN_INCLUDE_DIR})
+    endif()
+  endif()
+
+  find_cereal(${CMAKE_BINARY_DIR})
+  if (NOT CEREAL_FOUND)
+    get_deps(https://github.com/USCiLab/cereal/archive/refs/tags/v1.3.2.tar.gz cereal cereal-1.3.2.tar.gz)
+    set(CEREAL_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    find_cereal(${CMAKE_BINARY_DIR})
+    if (CEREAL_FOUND)
+      set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${CEREAL_INCLUDE_DIR})
+    endif()
+  endif()
+
+  find_mlpack_internal(${CMAKE_BINARY_DIR})
+  if (NOT MLPACK_FOUND)
+    get_deps(https://www.mlpack.org/files/mlpack-latest.tar.gz mlpack mlpack-latest.tar.gz)
+    set(MLPACK_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    find_mlpack(${CMAKE_BINARY_DIR})
+    if (MLPACK_FOUND)
+      set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${MLPACK_INCLUDE_DIR})
+    endif()
+  endif()
+
+  find_openmp() 
+
+endmacro()
+
+##===================================================
+##  MLPACK MAIN FUNCTIONS CALL. 
+##===================================================
+
+macro(find_mlpack)
+  # If we're using gcc, then we need to link against pthreads to use std::thread,
+  # which we do in the tests.
+  if (CMAKE_COMPILER_IS_GNUCC)
+    find_package(Threads)
+    set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT})
+  endif()
+
+  if (NOT MLPACK_DISABLE_OPENMP)
+    find_openmp()
+  endif ()
+
+  find_armadillo()
+  if (ARMADILLO_FOUND)
+    set(MLPACK_INCLUDE_DIRS ${ARMADILLO_INCLUDE_DIRS})
+    set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${ARMADILLO_LIBRARIES})
+  else()
+    message(FATAL_ERROR "Armadillo not found, (required dependency of mlpack).")
+  endif ()
+
+  find_ensmallen()
+  if (ENSMALLEN_FOUND)
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${ENSMALLEN_INCLUDE_DIR})
+  else()
+    message(FATAL_ERROR "Ensmallen not found, (required dependency of mlpack).")
+  endif()
+
+  find_cereal()
+  if (CEREAL_FOUND)
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${CEREAL_INCLUDE_DIR})
+  else()
+    message(FATAL_ERROR "Cereal not found, (required dependency of mlpack).")
+  endif()
+
+  if (USE_SYSTEM_STB)
```

```suggestion
  if (MLPACK_USE_SYSTEM_STB)
```
Just convention for CMake scripts, I always see this with the package as the prefix of the name.  I also added documentation for this option.